File: day_3/main.cpp
#include <bits/stdc++.h>

using namespace std;

#define DEBUG(var) \
    (std::cout << "[DEBUG]: " << #var << " = " << (var) << '\n')

typedef long long ll;
typedef unsigned long long ull;

ll part1();
ll part2();

int main()
{
    cout << part1() << endl;
    cout << part2() << endl;
}

string max_joltage_helper(map<pair<int, int>, string> &dp, const string &bank, int start, int depth)
{
    if (depth == 0 || start >= bank.length() || depth > bank.length() - start)
        return "";

    if (dp.find({start, depth}) != dp.end())
        return dp[{start, depth}];

    dp[{start, depth}] = max(
        bank[start] + max_joltage_helper(dp, bank, start + 1, depth - 1),
        max_joltage_helper(dp, bank, start + 1, depth));

    return dp[{start, depth}];
}

string max_joltage(const string &bank, int depth)
{
    map<pair<int, int>, string> dp;
    return max_joltage_helper(dp, bank, 0, depth);
}

ll part1()
{
    ifstream input("../data/3.txt");
    string line;
    ll res = 0;

    while (getline(input, line))
    {
        string a = max_joltage(line, 2);
        // DEBUG(a);
        res += stoll(a);
    }

    return res;
}

ll part2()
{
    ifstream input("../data/3.txt");
    string line;
    ll res = 0;

    while (getline(input, line))
    {
        string a = max_joltage(line, 12);
        // DEBUG(a);
        res += stoll(a);
    }

    return res;
}

File: day_3/main.py
def max_joltage(bank: str, depth: int = 12):
    dp = {}

    def helper(bank, depth):
        if len(bank) == 0 or depth > len(bank) or depth == 0:
            return ""

        if (bank, depth) in dp:
            return dp[(bank, depth)]

        dp[(bank, depth)] = max(
            bank[0] + helper(bank[1:], depth - 1), (helper(bank[1:], depth))
        )

        return dp[(bank, depth)]

    return int(helper(bank, depth))


res = 0

f = open("../data/3.txt")
for line in f.readlines():
    res += max_joltage(line.strip(), 2)
f.close()
print(res)

res = 0
f = open("../data/3.txt")
for line in f.readlines():
    res += max_joltage(line.strip(), 12)

print(res)
f.close()
